File: packages/python/hushspec/middleware.py
from __future__ import annotations

import json
from typing import Callable, Optional, TYPE_CHECKING

from hushspec.evaluate import Decision, EvaluationAction, EvaluationResult, evaluate
from hushspec.parse import parse_or_raise
from hushspec.schema import HushSpec

if TYPE_CHECKING:
    from hushspec.observer import EvaluationObserver

WarnHandler = Callable[[EvaluationResult, EvaluationAction], bool]


class HushSpecDenied(Exception):
    def __init__(self, result: EvaluationResult) -> None:
        self.result = result
        reason = result.reason or result.matched_rule or "policy denial"
        super().__init__(f"Action denied: {reason}")


class HushGuard:
    """Fail-closed policy guard: wraps evaluate / check / enforce semantics."""

    def __init__(
        self,
        policy: HushSpec,
        on_warn: Optional[WarnHandler] = None,
        observer: Optional["EvaluationObserver"] = None,
    ) -> None:
        self._policy = policy
        self._on_warn: WarnHandler = on_warn or (lambda _r, _a: False)
        self._observable_evaluator = None
        self._policy_hash: Optional[str] = None
        if observer is not None:
            from hushspec.observer import ObservableEvaluator
            from hushspec.receipt import compute_policy_hash
            self._observable_evaluator = ObservableEvaluator()
            self._observable_evaluator.add_observer(observer)
            self._policy_hash = compute_policy_hash(policy)
            self._observable_evaluator.notify_policy_loaded(policy.name, self._policy_hash)

    @classmethod
    def from_file(
        cls,
        path: str,
        on_warn: Optional[WarnHandler] = None,
        observer: Optional["EvaluationObserver"] = None,
    ) -> HushGuard:
        with open(path) as f:
            spec = parse_or_raise(f.read())
        return cls(spec, on_warn, observer=observer)

    @classmethod
    def from_yaml(
        cls,
        yaml_str: str,
        on_warn: Optional[WarnHandler] = None,
        observer: Optional["EvaluationObserver"] = None,
    ) -> HushGuard:
        spec = parse_or_raise(yaml_str)
        return cls(spec, on_warn, observer=observer)

    def evaluate(self, action: EvaluationAction) -> EvaluationResult:
        if self._observable_evaluator is not None:
            return self._observable_evaluator.evaluate(self._policy, action)
        return evaluate(self._policy, action)

    def check(self, action: EvaluationAction) -> bool:
        result = self.evaluate(action)
        if result.decision == Decision.ALLOW:
            return True
        if result.decision == Decision.WARN:
            return self._on_warn(result, action)
        return False

    def enforce(self, action: EvaluationAction) -> None:
        result = self.evaluate(action)
        if result.decision == Decision.DENY:
            raise HushSpecDenied(result)
        if result.decision == Decision.WARN and not self._on_warn(result, action):
            raise HushSpecDenied(result)

    @staticmethod
    def map_tool_call(
        tool_name: str,
        args: Optional[dict] = None,
    ) -> EvaluationAction:
        return EvaluationAction(
            type="tool_call",
            target=tool_name,
            args_size=len(json.dumps(args)) if args is not None else None,
        )

    @staticmethod
    def map_file_read(path: str) -> EvaluationAction:
        return EvaluationAction(type="file_read", target=path)

    @staticmethod
    def map_file_write(path: str, content: Optional[str] = None) -> EvaluationAction:
        return EvaluationAction(type="file_write", target=path, content=content)

    @staticmethod
    def map_egress(domain: str) -> EvaluationAction:
        return EvaluationAction(type="egress", target=domain)

    @staticmethod
    def map_shell_command(command: str) -> EvaluationAction:
        return EvaluationAction(type="shell_command", target=command)

    def swap_policy(self, new_policy: HushSpec) -> None:
        previous_hash = self._policy_hash
        self._policy = new_policy
        if self._observable_evaluator is not None:
            from hushspec.receipt import compute_policy_hash
            self._policy_hash = compute_policy_hash(new_policy)
            self._observable_evaluator.notify_policy_reloaded(
                new_policy.name,
                self._policy_hash,
                previous_hash,
            )

File: packages/hushspec/src/middleware.ts
import type { HushSpec } from './schema.js';
import type { EvaluationAction, EvaluationResult } from './evaluate.js';
import { evaluate } from './evaluate.js';
import { parse } from './parse.js';
import { readFileSync } from 'node:fs';
import type { PolicyProvider } from './policy-provider.js';
import type { EvaluationObserver } from './observer.js';
import { ObservableEvaluator } from './observer.js';
import { computePolicyHash } from './receipt.js';

export type WarnHandler = (result: EvaluationResult, action: EvaluationAction) => boolean;

/** Fail-closed: warn decisions without an onWarn handler are treated as deny. */
export class HushGuard {
  private policy: HushSpec;
  private onWarn: WarnHandler;
  private observableEvaluator: ObservableEvaluator | null = null;
  private policyHash: string | null = null;
  private provider: PolicyProvider | null = null;

  constructor(policy: HushSpec, options?: {
    onWarn?: WarnHandler;
    observer?: EvaluationObserver;
    provider?: PolicyProvider;
  }) {
    this.policy = policy;
    this.onWarn = options?.onWarn ?? (() => false);
    this.provider = options?.provider ?? null;
    if (options?.observer) {
      this.observableEvaluator = new ObservableEvaluator();
      this.observableEvaluator.addObserver(options.observer);
      this.policyHash = computePolicyHash(policy);
      this.observableEvaluator.notifyPolicyLoaded(policy.name, this.policyHash);
    }
  }

  static fromFile(path: string, options?: { onWarn?: WarnHandler }): HushGuard {
    const content = readFileSync(path, 'utf8');
    const result = parse(content);
    if (!result.ok) {
      throw new Error(`Failed to parse policy: ${result.error}`);
    }
    return new HushGuard(result.value, options);
  }

  static fromYaml(yaml: string, options?: { onWarn?: WarnHandler }): HushGuard {
    const result = parse(yaml);
    if (!result.ok) {
      throw new Error(`Failed to parse policy: ${result.error}`);
    }
    return new HushGuard(result.value, options);
  }

  static async fromProvider(
    provider: PolicyProvider,
    options?: { onWarn?: WarnHandler },
  ): Promise<HushGuard> {
    const spec = await provider.load();
    const guard = new HushGuard(spec, { ...options, provider });
    provider.watch((newSpec) => guard.swapPolicy(newSpec));
    return guard;
  }

  evaluate(action: EvaluationAction): EvaluationResult {
    const policy = this.activePolicyResult();
    if ('decision' in policy) {
      return policy;
    }
    if (this.observableEvaluator) {
      return this.observableEvaluator.evaluate(policy, action);
    }
    return evaluate(policy, action);
  }

  check(action: EvaluationAction): boolean {
    const result = this.evaluate(action);
    if (result.decision === 'allow') return true;
    if (result.decision === 'warn') return this.onWarn(result, action);
    return false;
  }

  enforce(action: EvaluationAction): void {
    const result = this.evaluate(action);
    if (result.decision === 'deny') {
      throw new HushSpecDenied(result);
    }
    if (result.decision === 'warn' && !this.onWarn(result, action)) {
      throw new HushSpecDenied(result);
    }
  }

  static mapToolCall(toolName: string, args?: Record<string, unknown>): EvaluationAction {
    return {
      type: 'tool_call',
      target: toolName,
      args_size: args ? JSON.stringify(args).length : undefined,
    };
  }

  static mapFileRead(path: string): EvaluationAction {
    return { type: 'file_read', target: path };
  }

  static mapFileWrite(path: string, content?: string): EvaluationAction {
    return { type: 'file_write', target: path, content };
  }

  static mapEgress(domain: string): EvaluationAction {
    return { type: 'egress', target: domain };
  }

  static mapShellCommand(command: string): EvaluationAction {
    return { type: 'shell_command', target: command };
  }

  swapPolicy(newPolicy: HushSpec): void {
    const previousHash = this.policyHash;
    this.policy = newPolicy;
    if (this.observableEvaluator) {
      this.policyHash = computePolicyHash(newPolicy);
      this.observableEvaluator.notifyPolicyReloaded(
        newPolicy.name,
        this.policyHash,
        previousHash ?? undefined,
      );
    }
  }

  private activePolicyResult(): HushSpec | EvaluationResult {
    if (this.provider == null) {
      return this.policy;
    }

    try {
      const current = this.provider.current();
      if (current == null) {
        return {
          decision: 'deny',
          matched_rule: '__hushspec_policy_provider__',
          reason: 'policy provider has not loaded a policy yet',
        };
      }
      this.policy = current;
      return current;
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return {
        decision: 'deny',
        matched_rule: '__hushspec_policy_provider__',
        reason: `policy provider unavailable: ${message}`,
      };
    }
  }
}

export class HushSpecDenied extends Error {
  public readonly result: EvaluationResult;

  constructor(result: EvaluationResult) {
    super(`Action denied: ${result.reason ?? result.matched_rule ?? 'policy denial'}`);
    this.name = 'HushSpecDenied';
    this.result = result;
  }
}
